### ISSUE TITLE
weakref.finalize called when kubernetes already cleaned up the key_file

### ISSUE THREAD

**Issue by alippai** [dask/dask-kubernetes#281]
When a python app terminates, https://github.com/dask/dask-kubernetes/blob/ccb9864d463a3e14988af53a947a658cf32e7e10/dask_kubernetes/core.py#L690 this finalize() is called.
Here we construct a new kubernetes client, however if it uses a key file in `/tmp/key_file` format (it creates a temp file "extract" for the urllib3 package, because it's embedded in the kube config file) then it's not available at this point anymore. It was cleaned up when kubernetes module run it's shutdown procedure. This raises exceptions.

How can we somehow call the `finalize` manually before the script terminates (and check in the actual finalize call whether it was run or not)? I couldn't trigger this code path from the script :/
`del client`, `del cluster`, `gc.collect()`, `client.close()`, `cluster.close()`, `cluster.scheduler.close()` didn't work, I couldn't get rid if the instance (which can imply a memory leak too)

**Comment by alippai** [comment]
tldr; I'd like to do:
```python
cluster = KubeCluster.from_yaml('worker-spec.yml')
cluster.close()
# or
await cluster.close()
```
in a way that I know that `cluster` finished all the communications.
`.close()` would return only when the teardown succeeded (or failed)

**Comment by jacobtomlinson** [comment]
Thanks for raising this, and agreed.

We try to clean up resources with `finalize` to help folks who forget to clean up their cluster. However it looks like we may have lost the ability to do this regularly.

We construct a new client because `finalize` is called after the asyncio event loop is closed, so no async code will work. So we have to construct a new sync client.

My preference here would be to implement `cluster.close()` correctly. But also maintain a codepath to try and clean up on `finalize` if it hasn't been done manually.

One benefit of having `.close` implemented correctly is we can also use a context manager. Today this would work, but fall back to `finalize` to clean up.

```python
from dask.distributed import Client
from dask_kubernetes import KubeCluster

with KubeCluster.from_yaml('worker-spec.yml') as cluster:
    with Client(cluster) as client:
        # Do some Dask work
```

**Comment by alippai** [comment]
I even tried to add:
```python
while not len(gc.get_referrers(cluster)) == 1:
	sleep(0.1)
```
which somehow helps (after doing all the mentioned cleanup work above), but it calls into async code even that way. It must be some tornado magic (an old event/lambda firing) or IDK. It's definitely looks like a big task, I appreciate that you even consider cleaning up this.

**Comment by alippai** [comment]
Similarly I think `cluster.scale(n)` returns before the worker registration is complete. I'd start this job above with removing all background / orphanned tasks. That's pretty confusing from a dev perspective too (beside the unwanted side-effects). Should I file a separate issue for this?

**Comment by jacobtomlinson** [comment]
> Similarly I think cluster.scale(n) returns before the worker registration is complete. 

This is intentional. We want Dask to start work ASAP and not necessarily wait for all workers.

When running in sync mode if you wish to wait for a certain number of workers you can use [`Client.wait_for_workers`](https://distributed.dask.org/en/latest/api.html#distributed.Client.wait_for_workers).

If using an async cluster you can `await` the cluster object after scaling to block until registration is complete.

**Comment by alippai** [comment]
So this is likely handled in `.close()` and it's not causing traffic at a later point. Thanks.

**Comment by jacobtomlinson** [comment]
Closing due to inactivity
